UnityProject/Assets/Scripts/PlayGroups/Input/Triggers/PickUpTrigger.cs

```diff
@@ -38,27 +38,53 @@ public override void Interact(GameObject originator, Vector3 position, string ha
 			else
 			{
 				//Server actions
-				if (ValidatePickUp(originator, hand))
+				if ( !ValidatePickUp(originator, hand) )
 				{
-					GetComponent<RegisterItem>().Unregister();
+					//Rollback prediction (inform player about item's true state)
+					GetComponent<CustomNetTransform>().NotifyPlayer(originator);
 				}
 			}
 		}
 
 		[Server]
 		public bool ValidatePickUp(GameObject originator, string handSlot = null)
 		{
-			PlayerScript ps = originator.GetComponent<PlayerScript>();
-			string slotName = handSlot ?? UIManager.Hands.CurrentSlot.eventName;
-			if (PlayerManager.PlayerScript == null || !ps.playerNetworkActions.Inventory.ContainsKey(slotName))
+			var ps = originator.GetComponent<PlayerScript>();
+			var slotName = handSlot ?? UIManager.Hands.CurrentSlot.eventName;
+			var cnt = GetComponent<CustomNetTransform>();
+			var state = cnt.State;
+			if ( SlotUnavailable(ps, slotName) )
 			{
 				return false;
 			}
+			if ( cnt.IsFloating() ? !CanReachFloating(ps, state) : !ps.IsInReach(state.position) )
+			{
+				Debug.LogWarningFormat($"Not in reach! server pos:{state.position} player pos:{originator.transform.position} (floating={cnt.IsFloating()})");
+				return false;
+			}
+
+//			Debug.LogFormat($"Pickup success! server pos:{state.position} player pos:{originator.transform.position} (floating={cnt.IsFloating()})");
```

Comment:
Keeping this one in is fine, considering the complexity....

Reply:
It's only really useful for debugging